libraries/chain/webassembly/runtimes/eos-vm.cpp

```diff
@@ -129,48 +130,90 @@ class eos_vm_instantiated_module : public wasm_instantiated_module_interface {
-         // set wasm allocator per apply data
-         _runtime->_bkend.set_wasm_allocator(&context.control.get_wasm_allocator());
+         auto& bkend      = _runtime->_bkend;
+         auto& exec_ctx   = _runtime->_exec_ctx;
+         auto& wasm_alloc = context.control.get_wasm_allocator();
 
-         apply_options opts;
-         if(context.control.is_builtin_activated(builtin_protocol_feature_t::configurable_wasm_limits)) {
-            const wasm_config& config = context.control.get_global_properties().wasm_configuration;
-            opts = {config.max_pages, config.max_call_depth};
-         }
+         apply_options opts = get_apply_options(context);
          auto fn = [&]() {
             eosio::chain::webassembly::interface iface(context);
-            _runtime->_bkend.initialize(&iface, opts);
-            _runtime->_bkend.call(
+            bkend.initialize(&iface, opts);
+            bkend.call(
                 iface, "env", "apply",
                 context.get_receiver().to_uint64_t(),
                 context.get_action().account.to_uint64_t(),
                 context.get_action().name.to_uint64_t());
          };
+
+         execute(context, bkend, exec_ctx, wasm_alloc, fn, false);
+      }
+
+   private:
+      void execute(apply_context& context, backend_t& bkend, eos_vm_runtime<Impl>::context_t& exec_ctx, vm::wasm_allocator& wasm_alloc, std::function<void()> fn, bool multi_expr_callbacks_allowed) {
+         // set up backend to share the compiled mod in the instantiated
+         // module of the contract
+         bkend.share(*_instantiated_module);
+         // set exec ctx's mod to instantiated module's mod
+         exec_ctx.set_module(&(_instantiated_module->get_module()));
+         // link exe ctx to backend
+         bkend.set_context(&exec_ctx);
+         // set max_call_depth and max_pages to original values
+         bkend.reset_max_call_depth();
+         bkend.reset_max_pages();
+         // set wasm allocator per apply data
+         bkend.set_wasm_allocator(&wasm_alloc);
+
          try {
-            checktime_watchdog wd(context.trx_context.transaction_timer);
-            _runtime->_bkend.timed_run(std::move(wd), std::move(fn));
+            checktime_watchdog wd(context.trx_context.transaction_timer, multi_expr_callbacks_allowed);
+            bkend.timed_run(std::move(wd), std::move(fn));
          } catch(eosio::vm::timeout_exception&) {
             context.trx_context.checktime();
          } catch(eosio::vm::wasm_memory_exception& e) {
             FC_THROW_EXCEPTION(wasm_execution_error, "access violation: ${d}", ("d", e.detail()));
          } catch(eosio::vm::exception& e) {
             FC_THROW_EXCEPTION(wasm_execution_error, "eos-vm system failure: ${d}", ("d", e.detail()));
          }
+
+         if (multi_expr_callbacks_allowed) {
+            context.trx_context.checktime(); // protect against the case where during the removal of the callback, the timer expires.
```

Comment:
I haven't thought it through entirely but my hunch is this check needs to be higher up (like maybe at the end of `execute_sync_call()` or such). Think about a case where the callee is being run by OC but the caller by JIT. We would need to add this checktime in both OC & JIT; or preferably just one common place higher up.

Reply:
I moved it to execute_sync_call for now and added a comment to the OC issue. We will find a common place when doing OC change.

https://github.com/AntelopeIO/spring/pull/1251/commits/fc220492545a08e21482a7aebf9fcd6fddf993a3